File: go/topics_group_iterator.go
package turbodata

import (
	"bytes"
	"io"
)

type TopicsGroupIterator struct {
	rs io.ReadSeeker

	// Memory buffers.
	loadBuf     *ReusableBuffer
	IndexBuf    *ReusableBuffer
	messageBuf  *ReusableBuffer
	bytesReader *bytes.Reader

	// Sort/filter scratch + outputs (reused per chunk).
	scratch                *sortAndFilterMergeScratch
	filteredMessageIndexes []*messageIndexWithTopicId
	filteredMessageLens    []int64

	// Read config.
	topicIds        map[uint16]struct{}
	startTimestamp  int64
	endTimestamp    int64
	order           Order
	isCompressed    bool
	incrementFactor int

	// Index chunk queue.
	indexChunkInfoList         []*IndexChunkInfo
	indexChunkInfoLens         []int64
	currentIndexChunkInfoIndex int

	// Message index cursor.
	currentMessageIndex int
}

func newTopicsGroupIterator(it *MessageIterator, topicIds map[uint16]struct{}, topicsInfo *TopicsInfo) *TopicsGroupIterator {
	indexChunkInfoList := make([]*IndexChunkInfo, 0, len(topicsInfo.IndexChunkInfoList))
	indexChunkInfoLens := make([]int64, 0, len(topicsInfo.IndexChunkInfoList))
	for i, indexChunkInfo := range topicsInfo.IndexChunkInfoList {
		if indexChunkInfo.EndTimestamp < it.startTimestamp {
			continue
		}

		if indexChunkInfo.StartTimestamp > it.endTimestamp {
			break
		}

		indexChunkInfoList = append(indexChunkInfoList, indexChunkInfo)
		if i < len(topicsInfo.IndexChunkInfoList)-1 {
			indexChunkInfoLens = append(indexChunkInfoLens, topicsInfo.IndexChunkInfoList[i+1].Offset-indexChunkInfo.Offset)
		} else {
			indexChunkInfoLens = append(indexChunkInfoLens, topicsInfo.TotalLen-indexChunkInfo.Offset+topicsInfo.IndexChunkInfoList[0].Offset)
		}
	}

	isCompressed, ok := topicsInfo.TopicMetadatas[0].Metadata["is_compressed"].(bool)
	if !ok {
		isCompressed = false
	}

	incrementFactor := 1
	currentIndexChunkInfoIndex := 0
	if it.order == ReverseTimeOrder {
		incrementFactor = -1
		currentIndexChunkInfoIndex = len(indexChunkInfoList) - 1
	}

	return &TopicsGroupIterator{
		rs: it.rs,

		loadBuf:     NewReusableBuffer(),
		IndexBuf:    NewReusableBuffer(),
		messageBuf:  NewReusableBuffer(),
		bytesReader: bytes.NewReader(nil),

		scratch:                newSortAndFilterMergeScratch(),
		filteredMessageIndexes: make([]*messageIndexWithTopicId, 0),
		filteredMessageLens:    make([]int64, 0),

		topicIds:        topicIds,
		startTimestamp:  it.startTimestamp,
		endTimestamp:    it.endTimestamp,
		order:           it.order,
		isCompressed:    isCompressed,
		incrementFactor: incrementFactor,

		indexChunkInfoList:         indexChunkInfoList,
		currentIndexChunkInfoIndex: currentIndexChunkInfoIndex,
		indexChunkInfoLens:         indexChunkInfoLens,
	}
}

func (it *TopicsGroupIterator) Next() (int64, uint16, []byte, error) {
	for it.currentMessageIndex >= len(it.filteredMessageIndexes) || it.currentMessageIndex < 0 {
		if it.currentIndexChunkInfoIndex >= len(it.indexChunkInfoList) || it.currentIndexChunkInfoIndex < 0 {
			return 0, 0, nil, io.EOF
		}

		indexChunk, err := it.loadIndexChunk(it.indexChunkInfoList[it.currentIndexChunkInfoIndex], it.indexChunkInfoLens[it.currentIndexChunkInfoIndex])
		if err != nil {
			return 0, 0, nil, err
		}
		it.currentIndexChunkInfoIndex += it.incrementFactor

		sortAndFilterMerge(
			indexChunk.TopicIndexes,
			indexChunk.UncompressedLen,
			it.topicIds,
			it.startTimestamp,
			it.endTimestamp,
			it.scratch,
			&it.filteredMessageIndexes,
			&it.filteredMessageLens,
		)

		it.currentMessageIndex = 0
		if it.order == ReverseTimeOrder {
			it.currentMessageIndex = len(it.filteredMessageIndexes) - 1
		}

		if len(it.filteredMessageIndexes) > 0 {
			if err := it.loadDataChunk(indexChunk.ChunkOffset, indexChunk.ChunkLen); err != nil {
				return 0, 0, nil, err
			}
		}
	}

	messageIndex := it.filteredMessageIndexes[it.currentMessageIndex]
	rangeEnd := messageIndex.messageIndex.OffsetInChunk + it.filteredMessageLens[it.currentMessageIndex]
	it.currentMessageIndex += it.incrementFactor

	return messageIndex.messageIndex.Timestamp, messageIndex.topicId, it.messageBuf.Data[messageIndex.messageIndex.OffsetInChunk:rangeEnd], nil
}

func (it *TopicsGroupIterator) loadIndexChunk(info *IndexChunkInfo, len int64) (*IndexChunk, error) {
	if _, err := it.rs.Seek(info.Offset, io.SeekStart); err != nil {
		return nil, err
	}
	it.loadBuf.Prepare(int(len))

	if _, err := io.ReadFull(it.rs, it.loadBuf.Data); err != nil {
		return nil, err
	}

	if err := decompressInto(it.loadBuf.Data, it.IndexBuf); err != nil {
		return nil, err
	}

	it.bytesReader.Reset(it.IndexBuf.Data)
	return ReadIndexChunk(it.bytesReader)
}

func (it *TopicsGroupIterator) loadDataChunk(offset int64, len int64) error {
	if _, err := it.rs.Seek(offset, io.SeekStart); err != nil {
		return err
	}

	if it.isCompressed {
		it.loadBuf.Prepare(int(len))
		if _, err := io.ReadFull(it.rs, it.loadBuf.Data); err != nil {
			return err
		}

		if err := decompressInto(it.loadBuf.Data, it.messageBuf); err != nil {
			return err
		}
		return nil
	}

	it.messageBuf.Prepare(int(len))
	if _, err := io.ReadFull(it.rs, it.messageBuf.Data); err != nil {
		return err
	}

	return nil
}

File: ts/src/iterators/topics_group_iterator.ts
// Copyright 2026 Wanjia He
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Default-path per-group iterator. Lazily reads one index chunk and one data
// chunk at a time; mirrors go/topics_group_iterator.go but uses readAt
// (offset/length) instead of Read+Seek state.

import type { Decompressor } from "../compression.js";
import { BinaryReader, readIndexChunk } from "../io.js";
import type { ReadSource } from "../read_source.js";
import { sortAndFilter, type MessageRef } from "../sort_and_filter.js";
import { META_KEY_COMPRESSED, type IndexChunkInfo, type TopicsInfo } from "../types.js";

export class TopicsGroupIterator {
  private readonly rs: ReadSource;
  private readonly decompress: Decompressor;
  private readonly topicIds: Set<number>;
  private readonly startTimestamp: bigint;
  private readonly endTimestamp: bigint;
  private readonly reverse: boolean;
  private readonly isCompressed: boolean;
  private readonly videoDecodable: boolean;

  // Per-chunk metadata.
  private readonly indexChunkInfoList: IndexChunkInfo[];
  private readonly indexChunkInfoLens: bigint[];

  // Iteration cursors.
  private chunkIdx: number;
  private msgIdx: number;
  private currentMessages: MessageRef[] = [];
  private currentLens: bigint[] = [];

  // Most-recently-loaded data chunk (uncompressed bytes).
  private dataBuffer: Uint8Array | undefined;

  constructor(args: {
    rs: ReadSource;
    decompress: Decompressor;
    topicIds: Set<number>;
    startTimestamp: bigint;
    endTimestamp: bigint;
    reverse: boolean;
    topicsInfo: TopicsInfo;
    /** When true, snap the per-chunk lower bound back to the GOP key frame. */
    videoDecodable?: boolean;
  }) {
    this.rs = args.rs;
    this.decompress = args.decompress;
    this.topicIds = args.topicIds;
    this.startTimestamp = args.startTimestamp;
    this.endTimestamp = args.endTimestamp;
    this.reverse = args.reverse;
    this.videoDecodable = args.videoDecodable === true;

    const firstMeta = args.topicsInfo.topicMetadatas[0];
    const flag = firstMeta?.metadata.get(META_KEY_COMPRESSED);
    this.isCompressed = flag === true;

    // Time-range filter at the chunk level + compute compressed lengths.
    const filteredInfos: IndexChunkInfo[] = [];
    const filteredLens: bigint[] = [];
    const all = args.topicsInfo.indexChunkInfoList;
    for (let i = 0; i < all.length; i++) {
      const info = all[i]!;
      if (info.endTimestamp < args.startTimestamp) {
        continue;
      }
      if (info.startTimestamp > args.endTimestamp) {
        break;
      }
      filteredInfos.push(info);
      let ln: bigint;
      if (i < all.length - 1) {
        ln = all[i + 1]!.offset - info.offset;
      } else {
        // Last index chunk: its length is (group total) - (its offset) + (first chunk's offset).
        ln = args.topicsInfo.totalLen - info.offset + all[0]!.offset;
      }
      filteredLens.push(ln);
    }
    this.indexChunkInfoList = filteredInfos;
    this.indexChunkInfoLens = filteredLens;

    if (args.reverse) {
      this.chunkIdx = filteredInfos.length - 1;
    } else {
      this.chunkIdx = 0;
    }
    this.msgIdx = 0;
  }

  /** True if we have any in-scope index chunks. */
  hasAny(): boolean {
    return this.indexChunkInfoList.length > 0;
  }

  /**
   * Next emits one (timestamp, topicId, data) record or returns undefined on
   * EOF. The returned Uint8Array aliases an internal buffer; the caller must
   * copy if persistence is needed.
   */
  async next(): Promise<{
    timestamp: bigint;
    topicId: number;
    data: Uint8Array;
  } | undefined> {
    while (
      this.msgIdx >= this.currentMessages.length ||
      this.msgIdx < 0
    ) {
      if (
        this.chunkIdx >= this.indexChunkInfoList.length ||
        this.chunkIdx < 0
      ) {
        return undefined;
      }
      await this.loadCurrentChunk();
      this.chunkIdx += this.reverse ? -1 : 1;
      if (this.reverse) {
        this.msgIdx = this.currentMessages.length - 1;
      } else {
        this.msgIdx = 0;
      }
    }

    const ref = this.currentMessages[this.msgIdx]!;
    const len = this.currentLens[this.msgIdx]!;
    this.msgIdx += this.reverse ? -1 : 1;

    const start = Number(ref.offsetInChunk);
    const end = start + Number(len);
    const data = this.dataBuffer!.subarray(start, end);
    return { timestamp: ref.timestamp, topicId: ref.topicId, data };
  }

  private async loadCurrentChunk(): Promise<void> {
    const info = this.indexChunkInfoList[this.chunkIdx]!;
    const ln = this.indexChunkInfoLens[this.chunkIdx]!;

    // Read + decompress the index chunk.
    const compressedIndex = await this.rs.read(info.offset, ln);
    const indexBytes = this.decompress(compressedIndex);
    const indexChunk = readIndexChunk(new BinaryReader(indexBytes));

    const filtered = sortAndFilter(
      indexChunk.topicIndexes,
      indexChunk.uncompressedLen,
      this.topicIds,
      this.startTimestamp,
      this.endTimestamp,
      this.videoDecodable,
    );
    this.currentMessages = filtered.msgs;
    this.currentLens = filtered.lens;

    if (filtered.msgs.length === 0) {
      this.dataBuffer = undefined;
      return;
    }

    // Read (and possibly decompress) the data chunk.
    const dataBytes = await this.rs.read(
      indexChunk.chunkOffset,
      indexChunk.chunkLen,
    );
    this.dataBuffer = this.isCompressed ? this.decompress(dataBytes) : dataBytes;
  }
}
